osquery/core/init.cpp

```diff
@@ -758,17 +758,20 @@ void Initializer::requestShutdown(int retcode) {
     kExitCode = retcode;
   }
 
+  auto current_thread_id = std::this_thread::get_id();
+
   // Stop thrift services/clients/and their thread pools.
-  if (std::this_thread::get_id() != kMainThreadId &&
-      FLAGS_enable_signal_handler) {
+  if (current_thread_id != kMainThreadId && FLAGS_enable_signal_handler) {
     raise(SIGUSR1);
   } else {
     // The main thread is requesting a shutdown, meaning in almost every case
     // it is NOT waiting for a shutdown.
     // Exceptions include: tight request / wait in an exception handler or
     // custom signal handling.
     Dispatcher::stopServices();
-    waitForShutdown();
+
+    if (current_thread_id == kMainThreadId)
```

Comment:
Isn't this check superfluous? The else is only exercised if `!(current_thread_id != kMainThreadId)`

Reply:
The previous if doesn't only check for the thread id, but also if `FLAGS_enable_signal_handler` is enabled, which by default it's not.
This means that other threads can end up calling the else path.